rs/replicated_state/src/canister_state/queues.rs

```diff
@@ -165,6 +112,11 @@ pub struct NewCanisterQueues {
     /// senders is best effort.
     remote_subnet_input_schedule: VecDeque<CanisterId>,
 
+    /// Set of all canisters enqueued in either `local_subnet_input_schedule` or
+    /// `remote_subnet_input_schedule`, to ensure that a canister is scheduled at
+    /// most once.
+    input_schedule_canisters: BTreeSet<CanisterId>,
```

Comment:
Why do we need this field?

Reply:
In order to prevent a canister being enqueued multiple times in the input schedule if messages are enqueued, then all get dropped, then new messages get enqueued again.

The condition for adding something to an input schedule used to be `if input_queue.len() == 1`, but with message expiration/shedding in the picture, an input queue can repeatedly go thfough lengths 0, 1, 0, 1, 0, 1, ... without ever being scheduled.

I had to choose between something like this and dropping the invariant, including during checkpoint loading. The latter seemed potentially dangerous.

Reply:
I guess I meanwhile figured it out myself: when timing out/shedding doesn't touch the schedules it might leave an empty queue behind. So we remember the already scheduled senders to avoid scheduling it again when pushing the next message into the empty queue. Is that about right?